src/hipBin.cpp

```diff
@@ -0,0 +1,267 @@
+
+};
+
+
+// Implementation ================================================
+//===========================================================================
+
+HipBin::HipBin() {
+  hipBinNVPtr_ = new HipBinNvidia();
+  hipBinAMDPtr_ = new HipBinAmd();
+  bool platformDetected = false;
+  if (hipBinAMDPtr_->detectPlatform()) {
+    const PlatformInfo& platformInfo = hipBinAMDPtr_->getPlatformInfo();   // populates the struct with AMD info
+    platformVec_.push_back(platformInfo);
+    hipBinBasePtrs_.push_back(hipBinAMDPtr_);
+    platformDetected = true;
+  } else if (hipBinNVPtr_->detectPlatform()) {
+    const PlatformInfo& platformInfo = hipBinNVPtr_->getPlatformInfo();  // populates the struct with Nvidia info
+    platformVec_.push_back(platformInfo);
+    hipBinBasePtrs_.push_back(hipBinNVPtr_);
+    platformDetected = true;
+  }
+  // if no device is detected, then it is defaulted to AMD
+  if (!platformDetected) {
+    cout << "Device not supported - Defaulting to AMD" << endl;
+    const PlatformInfo& platformInfo = hipBinAMDPtr_->getPlatformInfo();   // populates the struct with AMD info
+    platformVec_.push_back(platformInfo);
+    hipBinBasePtrs_.push_back(hipBinAMDPtr_);
+  }
+}
+
+HipBin::~HipBin() {
+  delete hipBinNVPtr_;
+  delete hipBinAMDPtr_;
+  hipBinBasePtrs_.clear();  // clearing the vector so no one accesses the pointers
+  platformVec_.clear();     // clearing the platform vector as the pointers are deleted
+}
+
+
+// subtring is present in string
+bool HipBin::substringPresent(string fullString, string subString) const {
+  bool found = false;
+  if (fullString.find(subString) != std::string::npos) {
+    found = true;
+  }
+  return found;
+}
+
+vector<PlatformInfo>& HipBin::getPlaformInfo() {
+  return platformVec_;  // Return the populated platform info.
+}
+
+
+vector<HipBinBase*>& HipBin::getHipBinPtrs() {
+  return hipBinBasePtrs_;  // Return the populated device pointers.
+}
+
+
+void HipBin::executeHipCC(int argc, char* argv[]) {
+  cout << "HIPCC executable" <<endl;
+  vector<HipBinBase*>& platformPtrs = getHipBinPtrs();
+  std::vector<string> argvcc;
+  for (int i = 0; i < argc; i++) {
+    argvcc.push_back(argv[i]);
+  }
+  // 0th index points to the first platform detected.
+  // In the near future this vector will contain mulitple devices
+  platformPtrs.at(0)->executeHipCCCmd(argvcc);
+}
+
+
+void HipBin::executeHipConfig(int argc, char* argv[]) {
+  vector<HipBinBase*>& platformPtrs = getHipBinPtrs();
+  for (unsigned int j = 0; j < platformPtrs.size(); j++) {
+    if (argc == 1) {
+      platformPtrs.at(j)->printFull();
+    }
+    for (int i = 1; i < argc; ++i) {
+      HipBinCommand cmd;
+      cmd = gethipconfigCmd(argv[i]);
+      switch (cmd) {
+      case help: platformPtrs.at(j)->printUsage();
+        break;
+      case path: {
+        string hipPath;
+        hipPath = platformPtrs.at(j)->getHipPath();
+        cout << hipPath;
```

Comment:
```suggestion
        cout << platformPtrs[j]->getHipPath();
```
Similar change for the other cases in the below. No need to use at() as j is known to stay within the bounds of container.

Reply:
I prefer to use at as its more safe during a out of range indexing - see https://thispointer.com/c-how-to-get-element-by-index-in-vector-at-vs-operator/